File: seven-segment-view/src/main/java/com/bendaschel/sevensegmentview/SevenSegmentView.java
package com.bendaschel.sevensegmentview;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Point;
import android.graphics.RectF;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.view.View;

import java.util.Arrays;

public class SevenSegmentView extends View {

    private static final Path BASE_PATH = PathUtils.makePathFromPoints(Arrays.asList(
            new Point(4, 1),
            new Point(5, 0),
            new Point(4, -1),
            new Point(-4, -1),
            new Point(-5, 0),
            new Point(-4, 1)
    ));
    private static final String KEY_SAVE_STATE = "saved_current_value";
    private static final DIGITS DEFAULT_STATE = DIGITS.ALL_OFF;
    public static final String KEY_SUPER_STATE = "superState";

    private enum SEGMENTS {
        TOP(0, 0, 10),
        TOP_RIGHT(90, 5, 5),
        BOTTOM_RIGHT(90, 5, -5),
        BOTTOM(0, 0, -10),
        BOTTOM_LEFT(90, -5, -5),
        TOP_LEFT(90, -5, 5),
        CENTER(0, 0, 0);

        private final Path path;

        SEGMENTS(int rotate, int translateX, int translateY) {
            Matrix matrix = new Matrix();
            // Yes, this must be -Y because the graphics coord system is upside-down
            // from a normal cartesian plane
            matrix.setTranslate(translateX, -translateY);
            matrix.preRotate(rotate);
            path = new Path();
            BASE_PATH.transform(matrix, path);
        }
    }

    private enum DIGITS {
        ZERO(ON, ON, ON, ON, ON, ON, OFF),
        ONE(OFF, ON, ON, OFF, OFF, OFF, OFF),
        TWO(ON, ON, OFF, ON, ON, OFF, ON),
        THREE(ON, ON, ON, ON, OFF, OFF, ON),
        FOUR(OFF, ON, ON, OFF, OFF, ON, ON),
        FIVE(ON, OFF, ON, ON, OFF, ON, ON),
        SIX(ON, OFF, ON, ON, ON, ON, ON),
        SEVEN(ON, ON, ON, OFF, OFF, OFF, OFF),
        EIGHT(ON, ON, ON, ON, ON, ON, ON),
        NINE(ON, ON, ON, OFF, OFF, ON, ON),
        ALL_OFF(OFF, OFF, OFF, OFF, OFF, OFF, OFF);
        boolean [] activeSegments;
        DIGITS(boolean ... activeSegments) {
            this.activeSegments = activeSegments;
        }
    }

    private static final RectF SEGMENT_BOUNDS = calculateAllSegmentBounds();
    private int onColor;
    private int offColor;
    private static final boolean ON = true;
    private static final boolean OFF = false;

    private Matrix mCurrentTransformation;
    private Path mTemporaryPath;
    private int mCurrentValue = DIGITS.ALL_OFF.ordinal();
    private Paint mPaint;
    private boolean[] mActiveSegments;

    public SevenSegmentView(Context context) {
        this(context, null);
    }

    public SevenSegmentView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SevenSegmentView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.SevenSegmentView, 0, 0);

        offColor =  a.getColor(R.styleable.SevenSegmentView_offColor, Color.argb(50, Color.red(Color.RED), 0, 0));
        onColor = a.getColor(R.styleable.SevenSegmentView_onColor, Color.RED);

        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mCurrentTransformation = new Matrix();
        mTemporaryPath = new Path();
        mActiveSegments = DEFAULT_STATE.activeSegments;
    }

    public void setCurrentValue(int currentValue) {
        if (currentValue >=0 && currentValue <=9) {
            mActiveSegments = DIGITS.values()[currentValue].activeSegments;
        }else{
            mActiveSegments = DEFAULT_STATE.activeSegments;
        }
        mCurrentValue = currentValue;
        invalidate();
    }

    public int getCurrentValue(){
        return mCurrentValue;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        // https://stackoverflow.com/questions/12266899/onmeasure-custom-view-explanation
        int widthSize = MeasureSpec.getSize(widthMeasureSpec);
        int heightSize = MeasureSpec.getSize(heightMeasureSpec);
        float scaleFactor = calculateScaleFactor(widthSize, heightSize,
                SEGMENT_BOUNDS.width(), SEGMENT_BOUNDS.height());
        int desiredWidth = (int)(SEGMENT_BOUNDS.width() * scaleFactor);
        int desiredHeight = (int)(SEGMENT_BOUNDS.height() * scaleFactor);
        setMeasuredDimension(desiredWidth, desiredHeight);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        mCurrentTransformation.reset();
        int canvasWidth = canvas.getWidth();
        int canvasHeight = canvas.getHeight();
        mCurrentTransformation.setTranslate(canvasWidth / 2, canvasHeight / 2);
        // Scale properly to the aspect ratio of the canvas
        int drawHeight = canvasHeight - getPaddingBottom() - getPaddingTop();
        int drawWidth = canvasWidth - getPaddingLeft() - getPaddingRight();
        float scaleFactor = calculateScaleFactor(drawWidth, drawHeight,
                SEGMENT_BOUNDS.width(), SEGMENT_BOUNDS.height());
        mCurrentTransformation.preScale(scaleFactor, scaleFactor);

        for(SEGMENTS segment: SEGMENTS.values()) {
            segment.path.transform(mCurrentTransformation, mTemporaryPath);
            int position = segment.ordinal();
            mPaint.setColor(mActiveSegments[position] ? onColor : offColor);
            canvas.drawPath(mTemporaryPath, mPaint);
            mTemporaryPath.reset();
        }
    }

    private float calculateScaleFactor(float containerWidth, float containerHeight,
                                       float childWidth, float childHeight) {
        float heightRatio = containerHeight / childHeight;
        float widthRatio = containerWidth / childWidth;
        return Math.min(widthRatio, heightRatio);
    }

    private static RectF calculateAllSegmentBounds( ) {
        RectF outerBounds = new RectF();
        RectF segmentBounds = new RectF();
        SEGMENTS.TOP.path.computeBounds(segmentBounds, true);
        for (SEGMENTS segment: SEGMENTS.values()) {
            segment.path.computeBounds(segmentBounds, true);
            outerBounds.union(segmentBounds);
        }
        return outerBounds;
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        // https://stackoverflow.com/questions/3542333/how-to-prevent-custom-views-from-losing-state-across-screen-orientation-changes
        // Much less of a PITA than implementing a custom parcelable to hold one value;
        Bundle bundle = new Bundle();
        bundle.putParcelable(KEY_SUPER_STATE, super.onSaveInstanceState());
        bundle.putInt(KEY_SAVE_STATE, mCurrentValue);
        return bundle;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        if (state instanceof Bundle) {
            Bundle bundle = (Bundle) state;
            setCurrentValue(bundle.getInt(KEY_SAVE_STATE, DEFAULT_STATE.ordinal()));
            state = bundle.getParcelable(KEY_SUPER_STATE);
        }
        super.onRestoreInstanceState(state);
    }

    public int getOnColor() {
        return onColor;
    }

    public void setOnColor(int onColor) {
        this.onColor = onColor;
        invalidate();
    }

    public int getOffColor() {
        return offColor;
    }

    public void setOffColor(int offColor) {
        this.offColor = offColor;
        invalidate();
    }
}

File: seven-segment-view/src/main/java/com/bendaschel/sevensegmentview/SevenSegmentView.kt
package com.bendaschel.sevensegmentview

import android.content.Context
import android.graphics.*
import android.os.Bundle
import android.os.Parcelable
import android.util.AttributeSet
import android.view.View
import java.util.*

class SevenSegmentView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : View(context, attrs, defStyleAttr) {

    var onColor: Int = 0
        set(value) {
            field = value
            invalidate()
        }

    var offColor: Int = 0
        set(value) {
            field = value
            invalidate()
        }

    var currentValue = DIGITS.ALL_OFF.ordinal
        set(currentValue) {
            assert(currentValue in 0..9)
            field = currentValue
            invalidate()
        }

    private var mActiveSegments: BooleanArray = DEFAULT_STATE.activeSegments
        get() = if (currentValue in 0..9) {
            DIGITS.values()[currentValue].activeSegments
        } else {
            DEFAULT_STATE.activeSegments
        }

    private val mCurrentTransformation: Matrix
    private val mTemporaryPath: Path
    private val mPaint: Paint

    private enum class SEGMENTS private constructor(rotate: Int, translateX: Int, translateY: Int) {
        TOP(0, 0, 10),
        TOP_RIGHT(90, 5, 5),
        BOTTOM_RIGHT(90, 5, -5),
        BOTTOM(0, 0, -10),
        BOTTOM_LEFT(90, -5, -5),
        TOP_LEFT(90, -5, 5),
        CENTER(0, 0, 0);

        internal val path: Path

        init {
            val matrix = Matrix()
            // Yes, this must be -Y because the graphics coord system is upside-down
            // from a normal cartesian plane
            matrix.setTranslate(translateX.toFloat(), (-translateY).toFloat())
            matrix.preRotate(rotate.toFloat())
            path = Path()
            BASE_PATH.transform(matrix, path)
        }
    }

    private enum class DIGITS constructor(internal vararg var activeSegments: Boolean) {
        ZERO(ON, ON, ON, ON, ON, ON, OFF),
        ONE(OFF, ON, ON, OFF, OFF, OFF, OFF),
        TWO(ON, ON, OFF, ON, ON, OFF, ON),
        THREE(ON, ON, ON, ON, OFF, OFF, ON),
        FOUR(OFF, ON, ON, OFF, OFF, ON, ON),
        FIVE(ON, OFF, ON, ON, OFF, ON, ON),
        SIX(ON, OFF, ON, ON, ON, ON, ON),
        SEVEN(ON, ON, ON, OFF, OFF, OFF, OFF),
        EIGHT(ON, ON, ON, ON, ON, ON, ON),
        NINE(ON, ON, ON, OFF, OFF, ON, ON),
        ALL_OFF(OFF, OFF, OFF, OFF, OFF, OFF, OFF);
    }

    init {
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.SevenSegmentView, 0, 0)
        offColor = a.getColor(R.styleable.SevenSegmentView_offColor, Color.argb(50, Color.red(Color.RED), 0, 0))
        onColor = a.getColor(R.styleable.SevenSegmentView_onColor, Color.RED)
        mPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        mCurrentTransformation = Matrix()
        mTemporaryPath = Path()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        // https://stackoverflow.com/questions/12266899/onmeasure-custom-view-explanation
        val widthSize = View.MeasureSpec.getSize(widthMeasureSpec)
        val heightSize = View.MeasureSpec.getSize(heightMeasureSpec)
        val scaleFactor = calculateScaleFactor(widthSize.toFloat(), heightSize.toFloat(),
                SEGMENT_BOUNDS.width(), SEGMENT_BOUNDS.height())
        val desiredWidth = (SEGMENT_BOUNDS.width() * scaleFactor).toInt()
        val desiredHeight = (SEGMENT_BOUNDS.height() * scaleFactor).toInt()
        setMeasuredDimension(desiredWidth, desiredHeight)
    }

    override fun onDraw(canvas: Canvas) {
        mCurrentTransformation.reset()
        val canvasWidth = canvas.width
        val canvasHeight = canvas.height
        mCurrentTransformation.setTranslate((canvasWidth / 2).toFloat(), (canvasHeight / 2).toFloat())
        // Scale properly to the aspect ratio of the canvas
        val drawHeight = canvasHeight - paddingBottom - paddingTop
        val drawWidth = canvasWidth - paddingLeft - paddingRight
        val scaleFactor = calculateScaleFactor(drawWidth.toFloat(), drawHeight.toFloat(),
                SEGMENT_BOUNDS.width(), SEGMENT_BOUNDS.height())
        mCurrentTransformation.preScale(scaleFactor, scaleFactor)

        for (segment in SEGMENTS.values()) {
            segment.path.transform(mCurrentTransformation, mTemporaryPath)
            val position = segment.ordinal
            mPaint.color = if (mActiveSegments!![position]) onColor else offColor
            canvas.drawPath(mTemporaryPath, mPaint)
            mTemporaryPath.reset()
        }
    }

    private fun calculateScaleFactor(containerWidth: Float, containerHeight: Float,
                                     childWidth: Float, childHeight: Float): Float {
        val heightRatio = containerHeight / childHeight
        val widthRatio = containerWidth / childWidth
        return Math.min(widthRatio, heightRatio)
    }

    override fun onSaveInstanceState(): Parcelable? {
        // https://stackoverflow.com/questions/3542333/how-to-prevent-custom-views-from-losing-state-across-screen-orientation-changes
        // Much less of a PITA than implementing a custom parcelable to hold one value;
        val bundle = Bundle()
        bundle.putParcelable(KEY_SUPER_STATE, super.onSaveInstanceState())
        bundle.putInt(KEY_SAVE_STATE, currentValue)
        return bundle
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        var superState: Bundle? = null
        if (state is Bundle) {
            currentValue = state.getInt(KEY_SAVE_STATE, DEFAULT_STATE.ordinal)
            superState = state.getParcelable(KEY_SUPER_STATE)
        }
        super.onRestoreInstanceState(superState)
    }

    companion object {

        private val BASE_PATH = PathUtils.makePathFromPoints(Arrays.asList(
                Point(4, 1),
                Point(5, 0),
                Point(4, -1),
                Point(-4, -1),
                Point(-5, 0),
                Point(-4, 1)
        ))
        private val KEY_SAVE_STATE = "saved_current_value"
        private val DEFAULT_STATE = DIGITS.ALL_OFF
        val KEY_SUPER_STATE = "superState"

        private val SEGMENT_BOUNDS = calculateAllSegmentBounds()
        private val ON = true
        private val OFF = false

        private fun calculateAllSegmentBounds(): RectF {
            val outerBounds = RectF()
            val segmentBounds = RectF()
            SEGMENTS.TOP.path.computeBounds(segmentBounds, true)
            for (segment in SEGMENTS.values()) {
                segment.path.computeBounds(segmentBounds, true)
                outerBounds.union(segmentBounds)
            }
            return outerBounds
        }
    }
}
